src/index/VocabularyMerger.h

```diff
@@ -16,12 +17,106 @@
 #include "index/IndexBuilderTypes.h"
 #include "index/Vocabulary.h"
 #include "util/HashMap.h"
-#include "util/MmapVector.h"
 #include "util/ProgressBar.h"
+#include "util/Serializer/FileSerializer.h"
+#include "util/Serializer/SerializePair.h"
+#include "util/Serializer/SerializeVector.h"
 #include "util/TypeTraits.h"
 
-using IdPairMMapVec = ad_utility::MmapVector<std::pair<Id, Id>>;
-using IdPairMMapVecView = ad_utility::MmapVectorView<std::pair<Id, Id>>;
+// Writes pairs of (partial ID, global ID) incrementally to a file
+class IdMap {
+ private:
+  uint64_t count_ = 0;
+  std::string filename_;
+  std::unique_ptr<ad_utility::serialization::FileWriteSerializer> serializer_;
+  bool headerWritten_ = false;
+
+ public:
+  explicit IdMap(const std::string& filename) : filename_(filename) {
+    serializer_ =
+        std::make_unique<ad_utility::serialization::FileWriteSerializer>(
+            filename);
+  }
+
+  IdMap(size_t, const std::string& filename) : IdMap(filename) {}
+
+  void push_back(const std::pair<Id, Id>& pair) {
+    // Write header (vector size) on first push_back - we'll update it in
+    // destructor
+    if (!headerWritten_) {
+      // Reserve space for size, will be updated in destructor
+      uint64_t placeholder = 0;
+      *serializer_ << placeholder;
+      headerWritten_ = true;
+    }
+    // Write the pair directly to file
+    *serializer_ << pair;
+    ++count_;
+  }
+
+  ~IdMap() {
+    if (serializer_) {
+      try {
+        // Update the size at the beginning of the file
+        auto currentPos = serializer_->getSerializationPosition();
+        serializer_->setSerializationPosition(0);
+        *serializer_ << count_;
+        serializer_->setSerializationPosition(currentPos);
```

Comment:
I don't think you need to restore the position if the file is directly closed afterwards.